src/MPMCQueue.NET/MPMCQueue.cs

```diff
@@ -4,88 +4,113 @@
 
 namespace MPMCQueue.NET
 {
-    [StructLayout(LayoutKind.Explicit, Size = 192, CharSet = CharSet.Ansi)]
+    [StructLayout(LayoutKind.Explicit, Size = 336)]
     public class MPMCQueue
     {
-        [FieldOffset(0)]
-        private readonly Cell[] _buffer;
-        [FieldOffset(8)]
-        private readonly int _bufferMask;
-        [FieldOffset(64)]
-        private int _enqueuePos;
-        [FieldOffset(128)]
-        private int _dequeuePos;
+        /// <summary>
+        /// 128 bytes cache line already exists in some CPUs.
+        /// </summary>
+        /// <remarks>
+        /// Also "the spatial prefetcher strives to keep pairs of cache lines in the L2 cache."
+        /// https://stackoverflow.com/questions/29199779/false-sharing-and-128-byte-alignment-padding
+        /// </remarks>
+        internal const int SAFE_CACHE_LINE = 128;
 
+        [FieldOffset(SAFE_CACHE_LINE)]
+        private readonly Cell[] _enqueueBuffer;
+
+        [FieldOffset(SAFE_CACHE_LINE + 8)]
+        private volatile int _enqueuePos;
+
+        // Separate access to buffers from enqueue and dequeue.
+        // This removes false sharing and accessing a buffer 
+        // reference also prefetches the following Pos with [(64 - (8 + 4 + 4)) = 52]/64 probability.
+
+        [FieldOffset(SAFE_CACHE_LINE * 2)]
+        private readonly Cell[] _dequeueBuffer;
+
+        [FieldOffset(SAFE_CACHE_LINE * 2 + 8)]
+        private volatile int _dequeuePos;
 
         public MPMCQueue(int bufferSize)
         {
             if (bufferSize < 2) throw new ArgumentException($"{nameof(bufferSize)} should be greater than or equal to 2");
             if ((bufferSize & (bufferSize - 1)) != 0) throw new ArgumentException($"{nameof(bufferSize)} should be a power of 2");
 
-            _bufferMask = bufferSize - 1;
-            _buffer = new Cell[bufferSize];
+            _enqueueBuffer = new Cell[bufferSize];
 
             for (var i = 0; i < bufferSize; i++)
             {
-                _buffer[i] = new Cell(i, null);
+                _enqueueBuffer[i] = new Cell(i, null);
             }
 
+            _dequeueBuffer = _enqueueBuffer;
             _enqueuePos = 0;
             _dequeuePos = 0;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
This is a significant change by itself.

Is the per the same without AI?

Reply:
Without AI is slightly slower of cause. But the main point of changes in multi-threaded scalability: spinning and two references to the same buffer. Not sure about buffer bounds check with `& buffer.Length - 1` - does it work now? Am still lazy to learn assembly.

Reply:
The check is still there, but JIT reuses the load of Length for both the check and `buffer.Length - 1`.

Reply:
Is there any trick to avoid bounds check here? To leverage pow3 somehow. I searched through coreclr issues but couldn't find any. Also do you know if there is a single place with all current optimizations for BC elimination?

Reply:
I don't think it can avoid bounds check.
With `int index = pos & (buffer.Length - 1)` `pos` can be anything and `buffer.Length` can be zero.

Reply:
This is my bad habbit in general. But for low-level building blocks that are intended to be used later it's often better to have it: you could always wrap it into another method if you do not want inlining or prefer JIT to decide. Usually the caller decides this, not callee (relevant discussion https://github.com/dotnet/corefxlab/issues/2592#issuecomment-462515324). Here I never compared perf to the original implementation in absolute numbers, but the multithreaded scalability of the same method, which doesn't depend on the attribute.